### ISSUE TITLE
Cannot deploy to Azure function - Kudu gives 'username or password is incorrect'

### ISSUE THREAD

**Issue by TrainingIQ** [projectkudu/kudu#2759]
### Repro steps.
-Function  Url is https://tiqsubcriptions.azurewebsites.net. 
- Attempting to access https://tiqsubcriptions.scm.azurewebsites.net/ gives 500 Username or password is incorrect.
- Slots feature is enabled.

### The log/error given by the failure.

Server Error in '/' Application.
The user name or password is incorrect.
Description: An unhandled exception occurred during the execution of the current web request. Please review the stack trace for more information about the error and where it originated in the code. 

Exception Details: System.IO.IOException: The user name or password is incorrect.

Source Error: 

An unhandled exception was generated during the execution of the current web request. Information regarding the origin and location of the exception can be identified using the exception stack trace below.

Stack Trace: 

[IOException: The user name or password is incorrect.
]
   System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) +12701600
   System.IO.Directory.InternalCreateDirectory(String fullPath, String path, Object dirSecurityObj, Boolean checkHost) +12468710
   System.IO.Directory.InternalCreateDirectoryHelper(String path, Boolean checkHost) +82
   System.IO.Directory.CreateDirectory(String path) +33
   System.IO.Abstractions.DirectoryWrapper.CreateDirectory(String path) +11
   Kudu.Core.Infrastructure.FileSystemHelpers.CreateDirectory(String path) in C:\Kudu Files\Private\src\master\Kudu.Core\Infrastructure\FileSystemHelpers.cs:32
   Kudu.Core.Environment.get_DeploymentToolsPath() in C:\Kudu Files\Private\src\master\Kudu.Core\Environment.cs:183
   Kudu.Core.Infrastructure.PathUtilityBase.GetPathFolders(IEnvironment environment) in C:\Kudu Files\Private\src\master\Kudu.Core\Infrastructure\PathUtils\PathUtilityBase.cs:23
   Kudu.Services.Web.App_Start.NinjectServices.PrependFoldersToPath(IEnvironment environment) in C:\Kudu Files\Private\src\master\Kudu.Services.Web\App_Start\NinjectServices.cs:758
   Kudu.Services.Web.App_Start.NinjectServices.RegisterServices(IKernel kernel) in C:\Kudu Files\Private\src\master\Kudu.Services.Web\App_Start\NinjectServices.cs:152
   Kudu.Services.Web.App_Start.NinjectServices.CreateKernel() in C:\Kudu Files\Private\src\master\Kudu.Services.Web\App_Start\NinjectServices.cs:129
   Ninject.Web.Common.Bootstrapper.Initialize(Func`1 createKernelCallback) +16
   Kudu.Services.Web.App_Start.NinjectServices.Start() in C:\Kudu Files\Private\src\master\Kudu.Services.Web\App_Start\NinjectServices.cs:95

[TargetInvocationException: Exception has been thrown by the target of an invocation.]
   System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) +0
   System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) +79
   System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) +101
   WebActivatorEx.BaseActivationMethodAttribute.InvokeMethod() +73
   WebActivatorEx.ActivationManager.RunActivationMethods(Boolean designerMode) +637
   WebActivatorEx.ActivationManager.Run() +84

[InvalidOperationException: The pre-application start initialization method Run on type WebActivatorEx.ActivationManager threw an exception with the following error message: Exception has been thrown by the target of an invocation..]
   System.Web.Compilation.BuildManager.InvokePreStartInitMethodsCore(ICollection`1 methods, Func`1 setHostingEnvironmentCultures) +615
   System.Web.Compilation.BuildManager.InvokePreStartInitMethods(ICollection`1 methods) +141
   System.Web.Compilation.BuildManager.CallPreStartInitMethods(String preStartInitListPath, Boolean& isRefAssemblyLoaded) +102
   System.Web.Compilation.BuildManager.ExecutePreAppStart() +157
   System.Web.Hosting.HostingEnvironment.Initialize(ApplicationManager appManager, IApplicationHost appHost, IConfigMapPathFactory configMapPathFactory, HostingEnvironmentParameters hostingParameters, PolicyLevel policyLevel, Exception appDomainCreationException) +549

[HttpException (0x80004005): The pre-application start initialization method Run on type WebActivatorEx.ActivationManager threw an exception with the following error message: Exception has been thrown by the target of an invocation..]
   System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +10043436
   System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +95
   System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +254

### Mention any other details that might be useful.

I tried to deploy from Visual Studio 2017 today and got the "username or password is incorrect" error. The only thing I did relative to the previous successful deployment (24 hours ago) was add a new function to the project.

**Comment by davidebbo** [comment]
Most likely unrelated to Kudu. Check that your `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` has a valid storage connection string.

**Comment by marshallr12** [comment]
Yes it has a valid connection string.

**Comment by marshallr12** [comment]
I tried what was suggested here: https://github.com/Microsoft/vsts-tasks/issues/6513 . Setting WEBSITE_CONTENTSHARE = 0 doesn't fix the issue for me.

**Comment by davidebbo** [comment]
While not Kudu related, I do think there is a bug in how slots deal with Azure Files. What's slightly unusual in your scenario is that you are using a different storage account for the two slots. One uses trainingqigprod, and the other one trainingqigdev. And then each has a different content share name. Then I think when you swapped some wires got crossed and it ended up trying to use the storage account from one with the share name from the other, or something like that.

@ruslany can probably comment further.

**Comment by marshallr12** [comment]
I wonder if those settings should be set slotted? As a workaround, I'm creating two separate function apps - one for deployment and one for production without using the slots feature. But I'll leave existing app up if anyone wants to poke around. I will try swapping again and see if that helps.

**Comment by davidebbo** [comment]
No, slot settings don't make sense here since you really want your files to move back and forth as you swap (kind of the main point of swapping!).

There is a clear bug here, so until we figure it out, it's best to avoid slots with Function Apps (they are only in Preview BTW).

**Comment by davidebbo** [comment]
But yeah, please do keep the app around in case we can't repro.

**Comment by marshallr12** [comment]
What I found was WEBSITE_CONTENTSHARE wasn't slotted (not sure if, but after setting it to be slotted, and swapping again, things worked once more. So you're saying WEBSITE_CONTENTAZUREFILECONNECTIONSTRING shouldn't be slotted either?

**Comment by marshallr12** [comment]
Oh, but the staging slot isn't working now. However, prior to setting WEBSITE_CONTENTSHARE to be slotted, neither the master nor staging slot were working. Yeah, I'll avoid slots for now, but they would be very useful once in release.

**Comment by davidebbo** [comment]
Oh, you have `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` as slotted? How did it end up that way? That feels very wrong to me. If you slot both, it may look like it's not broken, but for all practical purpose, your swap will not work, since each slot will be stuck to the same files. Does that make sense? Does everything work if you unslot both (after starting from a working state)?

**Comment by TrainingIQ** [comment]
I'll try unslotting both and see what happens.

**Comment by TrainingIQ** [comment]
Unslotting both settings seemed to work. I believe I created the slot through the portal. But WEBSITE_CONTENTAZUREFILECONNECTIONSTRING was pointed to production, and I changed it to point to my development storage account (intuitively, that's where I want all  my development files to be). So you're saying all the slots should probably use the same storage account as the parent, regardless of their purpose?
